polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -220,7 +222,7 @@ void MLIRScanner::init(mlir::func::FuncOp func, const FunctionDecl *fd) {
     if (isArray) {
       params.emplace(parm, ValueCategory(val, /*isReference*/ true));
     } else {
-      auto alloc = createAllocOp(val.getType(), parm, /*memspace*/ 0, isArray,
+      auto alloc = createAllocOp(val.getType(), parm, defaultAddrSpace, isArray,
```

Comment:
Sorry I'm late to review this

So, in the SPIR convention, alloca must be in the address space `0` (private address space). Usage of `alloc` should have a check as to whether the operation except a pointer to private or to generic, in which case a cast is required.

In most cases, the AST should have a cast operation `CK_AddressSpaceConversion` which would drive that conversion. Of course there is local tweaks to have, I recall some around here had to fix the codegen to make things work properly.

Other point, the default address space in LLVM refers to the address space for the alloc (expressed in the datalayout string). The `defaultAddrSpace` you meant is usually called flat address space.

Reply:
Beyond the datalayout violation in the LLVM module, in SPIR-V this would lead to creation of `OpVariable` with a storage class to Generic which is invalid https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpVariable

Reply:
@Naghasan thanks for the comments. I was not aware that an alloca in addrspace(4) could not be translated to SPIRV! Indeed I verified that is the case by generating the LLVM IR for a simple test and then calling `llvm-spirv` translator on it:
```
llvm-spirv t.bc     
InvalidInstruction: Can't translate llvm instruction:
   %5 = alloca %"class.cl::sycl::range.1", i64 %4, align 8, addrspace(4)
```

Reply:
I will post a reversal of this PR.

Reply:
Reversal posted: https://github.com/intel/llvm/pull/6900